### DISCUSSION TITLE
Is it possible to write metadata into index files?

### DISCUSSION THREAD

**Discussion by xz328** [facebookresearch/faiss#3566] [General]
# Summary
Hi, 

I am trying to write metadata, e.g., the date, the user, the faiss version, and other dependencies into index files. Is there a way to achieve this?

**Comment by mdouze** [comment]
You can prepend or append it to the index data. `index_read` will just sequentially read the index data and leave the read pointer where at the beginning of the remaining data. See 

https://github.com/facebookresearch/faiss/blob/main/faiss/index_io.h#L39

**Comment by xz328** [comment]
> 
Thanks. Could you explain more about how I can **prepend** or **append** the metadata to the index data, i.e., using what method?

The mentioned line of the code (https://github.com/facebookresearch/faiss/blob/main/faiss/index_io.h#L39) seems to explain that the write_index method takes two arguments, one the index object and the other file name. I don't quite get how that relates to my question, as I am not a C programmer. Could you explain further if I have misunderstood it? 

Thanks

**Comment by MrCabss69** [comment]
LangChain's integration keeps the metadata in a separate structure that maps to the index items. Instead of trying to write the metadata inside the FAISS index file, you can store it in a separate file/database and maintain the relationship between the two using identifiers.

[Source Code](https://github.com/langchain-ai/langchain/blob/master/libs/community/langchain_community/vectorstores/faiss.py#L385)
